tasklog/log.go

```diff
@@ -214,6 +227,9 @@ func (l *Logger) logTask(task Task) {
 
 	var update *Update
 	for update = range task.Updates() {
+		if !isatty.IsTerminal(os.Stdout.Fd()) && !l.forceProgress {
```

Comment:
Does `isatty` cache this result? If not, and if it's expensive to compute, we should cache it ourselves, since this loop is expected to run many times.

On another note, I don't think that this is quite what we want: indeed this checks whether we can write to stdout, but stdout is not always the `io.Writer` we are writing to. Could you instead say:

```go
type Logger struct {
        ttyOnce sync.Once
        tty bool
}

func (l *Logger) tty() bool {
        l.ttyOnce.Do(func() {
                if f, ok := l.sink.(interface{ // <- could use `*os.File`, too 
                        Fd() uintptr
                }); ok {
                        l.tty = isatty.IsTerminal(f.Fd())
                        return
                }
        })
        return l.tty
}
```

And then rewrite this line to:

```go
if !(l.tty() || l.forceProgress) {
        continue
}
```

Reply:
Checking the (possibly non-existing) `Fd()` on the actual `Writer` makes good sense. Won't work in the cases where the `sink` is a `io.MultiWriter` (`command_fetch.go` and `command_prune.go` does this), but I'll try to come up with a solution to that. Maybe wrapping the `io.MultiWriter` in a type that exposes the `Fd()` of its first writer. 

The `sink` won't change during the lifetime of the logger, so could we not cache result of the call to `isatty` on instantiation of the logger?